Userland/Libraries/LibJS/Runtime/AbstractOperations.cpp

```diff
@@ -1316,4 +1317,159 @@ ThrowCompletionOr<String> get_substitution(VM& vm, Utf16View const& matched, Utf
+Completion dispose(VM& vm, Value value, NonnullGCPtr<FunctionObject> method)
+{
+    // 1. Let result be ? Call(method, V).
+    [[maybe_unused]] auto result = TRY(call(vm, *method, value));
+
+    // NOTE: Hint can only be sync-dispose so we ignore step 2.
+    // 2. If hint is async-dispose and result is not undefined, then
+    //    a. Perform ? Await(result).
+
+    // 3. Return undefined.
+    return js_undefined();
+}
+
+// 2.1.6 DisposeResources ( disposable, completion ), https://tc39.es/proposal-explicit-resource-management/#sec-disposeresources-disposable-completion-errors
+Completion dispose_resources(VM& vm, Vector<DisposableResource> const* disposable, Completion completion)
+{
+    // 1. If disposable is not undefined, then
+    if (disposable) {
+        // a. For each resource of disposable.[[DisposableResourceStack]], in reverse list order, do
+        for (auto const& resource : disposable->in_reverse()) {
+            // i. Let result be Dispose(resource.[[ResourceValue]], resource.[[Hint]], resource.[[DisposeMethod]]).
+            auto result = dispose(vm, resource.resource_value, resource.dispose_method);
+
+            // ii. If result.[[Type]] is throw, then
+            if (result.is_error()) {
+                // 1. If completion.[[Type]] is throw, then
+                if (completion.is_error()) {
+                    // a. Set result to result.[[Value]].
+
+                    // b. Let suppressed be completion.[[Value]].
+                    auto suppressed = completion.value().value();
+
+                    // c. Let error be a newly created SuppressedError object.
+                    auto error = SuppressedError::create(*vm.current_realm());
+
+                    // d. Perform ! DefinePropertyOrThrow(error, "error", PropertyDescriptor { [[Configurable]]: true, [[Enumerable]]: false, [[Writable]]: true, [[Value]]: result }).
+                    MUST(error->define_property_or_throw(vm.names.error, { .value = result.value(), .writable = true, .enumerable = true, .configurable = true }));
+
+                    // e. Perform ! DefinePropertyOrThrow(error, "suppressed", PropertyDescriptor { [[Configurable]]: true, [[Enumerable]]: false, [[Writable]]: true, [[Value]]: suppressed }).
+                    MUST(error->define_property_or_throw(vm.names.suppressed, { .value = suppressed, .writable = true, .enumerable = false, .configurable = true }));
+
+                    // f. Set completion to ThrowCompletion(error).
+                    completion = throw_completion(error);
+                }
+                // 2. Else,
+                else {
+                    // a. Set completion to result.
+                    completion = result;
+                }
+            }
+        }
+    }
+
+    // 2. Return completion.
+    return completion;
+}
+
+Completion dispose_resources(VM& vm, DeclarativeEnvironment const* disposable, Completion completion)
```

Comment:
Likewise re ptr -> optional

Reply:
Environments are generally used with pointers so wrapping them in optionals is not super nice at every callsite.

Just changing it to take a `Optional<DeclarativeEnvironment const&>` means that at every callsite where we have a pointer (all of them) we have to do:
```cpp
// At this point we should have already checked that if env is not 
//  null it is a declarative environment.
Optional<DeclarativeEnvironment const&> disposable;
if (env)
	env = *static_cast<DeclarativeEnvironment>(disposable);

result = dispose_resources(vm, disposable, result);
```

As I see it we have 3 options:
- Leave it as is (not ideal)
- Add pointer constructor or something to Optional to make this nice (I don't know the implications of this but just testing it seems fine)
- Add a bunch of optional noise at every call site (I don't like this option but it's fine)

Of course if I'm missing some easy way to create the optional do let me know! I could be missing something.

Reply:
Alright, but can we at least upgrade those raw pointers to (nullable) `GCPtr<T>`? :)

Reply:
Of course!